File: test/cpp/qps/worker.cc
//
//
// Copyright 2015 gRPC authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
//

#include <signal.h>

#include <chrono>
#include <thread>
#include <vector>

#include "absl/flags/flag.h"

#include <grpc/grpc.h>
#include <grpc/support/time.h>

#include "src/core/lib/debug/stats.h"
#include "src/core/lib/debug/stats_data.h"
#include "test/core/util/test_config.h"
#include "test/cpp/qps/qps_worker.h"
#include "test/cpp/util/test_config.h"
#include "test/cpp/util/test_credentials_provider.h"

ABSL_FLAG(int32_t, driver_port, 0, "Port for communication with driver");
ABSL_FLAG(int32_t, server_port, 0,
          "Port for operation as a server, if not specified by the server "
          "config message");
ABSL_FLAG(std::string, credential_type, grpc::testing::kInsecureCredentialsType,
          "Credential type for communication with driver");

static bool got_sigint = false;

static void sigint_handler(int /*x*/) { got_sigint = true; }

namespace grpc {
namespace testing {

std::vector<grpc::testing::Server*>* g_inproc_servers = nullptr;

static void RunServer() {
  QpsWorker worker(absl::GetFlag(FLAGS_driver_port),
                   absl::GetFlag(FLAGS_server_port),
                   absl::GetFlag(FLAGS_credential_type));

  while (!got_sigint && !worker.Done()) {
    gpr_sleep_until(gpr_time_add(gpr_now(GPR_CLOCK_REALTIME),
                                 gpr_time_from_millis(500, GPR_TIMESPAN)));
  }
}

}  // namespace testing
}  // namespace grpc

int main(int argc, char** argv) {
  grpc::testing::TestEnvironment env(&argc, argv);
  grpc::testing::InitTest(&argc, &argv, true);

  signal(SIGINT, sigint_handler);

  grpc::testing::RunServer();
  gpr_log(GPR_ERROR, "Global Stats:\n%s",
          StatsAsJson(grpc_core::global_stats().Collect().get()).c_str());
  return 0;
}

File: src/python/grpcio_tests/tests_aio/benchmark/worker.py
# Copyright 2020 The gRPC Authors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import argparse
import asyncio
import logging

from grpc.experimental import aio

from src.proto.grpc.testing import worker_service_pb2_grpc
from tests_aio.benchmark import worker_servicer


async def run_worker_server(port: int) -> None:
    server = aio.server()

    servicer = worker_servicer.WorkerServicer()
    worker_service_pb2_grpc.add_WorkerServiceServicer_to_server(
        servicer, server)

    server.add_insecure_port('[::]:{}'.format(port))

    await server.start()

    await servicer.wait_for_quit()
    await server.stop(None)


if __name__ == '__main__':
    logging.basicConfig(level=logging.DEBUG)
    parser = argparse.ArgumentParser(
        description='gRPC Python performance testing worker')
    parser.add_argument('--driver_port',
                        type=int,
                        dest='port',
                        help='The port the worker should listen on')
    parser.add_argument('--uvloop',
                        action='store_true',
                        help='Use uvloop or not')
    args = parser.parse_args()

    if args.uvloop:
        import uvloop
        asyncio.set_event_loop_policy(uvloop.EventLoopPolicy())
        loop = uvloop.new_event_loop()
        asyncio.set_event_loop(loop)

    asyncio.get_event_loop().run_until_complete(run_worker_server(args.port))
